### PR TITLE
feat: service MCP — expose workflow services as MCP tools

### PR DESCRIPTION
**Pull request by edurdias**
## Overview

Extends workflow services with optional MCP (Model Context Protocol) support. When MCP is enabled on a service, each workflow becomes 5 dynamically generated MCP tools — allowing AI agents to discover, execute, resume, and monitor workflows through the standard MCP protocol.

Services serve both HTTP and MCP on the same process and port. The MCP tool info endpoint lives at `/mcp` (standalone) or `/services/{name}/mcp/tools` (built-in). Each new MCP client connection gets the latest tool set — new workflows are picked up automatically.

The existing general-purpose MCP server (`flux start mcp`) is unchanged — it remains the full admin control plane with 31 tools. Service MCP is a focused, per-service interface that exposes only the workflows an AI agent needs.

### Tool Generation (5 per workflow)

Each workflow in the service produces 5 MCP tools:

| Tool | Purpose |
|---|---|
| `{name}` | Run workflow synchronously, return output |
| `{name}_async` | Run asynchronously, return execution_id |
| `resume_{name}` | Resume paused workflow synchronously |
| `resume_{name}_async` | Resume paused workflow asynchronously |
| `status_{name}` | Check execution status |

For a service with `invoice` and `refund`, the agent sees 10 tools:

```
invoice(customer_id: str, amount: float)       ← typed! Pydantic extracted
invoice_async(customer_id: str, amount: float)
resume_invoice(execution_id: str, input: str)
resume_invoice_async(execution_id: str, input: str)
status_invoice(execution_id: str)

refund(input: str)                             ← generic, dict input
refund_async(input: str)
resume_refund(execution_id: str, input: str)
resume_refund_async(execution_id: str, input: str)
status_refund(execution_id: str)
```

### Typed Parameters from Pydantic Models

Input schemas are extracted **at workflow registration time** and stored in `WorkflowInfo.metadata`. When a workflow declares `ExecutionContext[InvoiceInput]` where `InvoiceInput` is a Pydantic BaseModel, the MCP tools get individual typed parameters. Otherwise, input is a generic JSON string.

Extraction runs in the same trust boundary as registration — the server already loads workflow source for the catalog. Schemas are persisted in the database and served via the `GET /services/{name}` API, so neither the proxy nor MCP server needs to load source code.

### `mcp_enabled` — Opt-In per Service

MCP is opt-in per service via a stored `mcp_enabled` flag:

```bash
flux service create billing --namespace billing --mcp     # MCP enabled at creation
flux service update billing --mcp                          # enable later
flux service update billing --no-mcp                       # disable
```

The standalone process reads the stored flag by default. CLI flags override:

```bash
flux service start billing --port 9000           # reads stored mcp_enabled
flux service start billing --port 9000 --mcp     # override: force enable
flux service start billing --port 9000 --no-mcp  # override: force disable
```

### Per-Session Tool Generation

Tools are generated once per MCP client connection from the current resolved endpoint set. Existing connections keep their tools. New workflows appear on the next client connection — no process restart needed.

### EndpointProvider Protocol

`ServiceMCPServer` accepts an `EndpointProvider` protocol, abstracting resolution:
- **Standalone:** `ProxyEndpointProvider` fetches endpoints from the Flux server via HTTP (cached with TTL)
- **Built-in:** Direct `ServiceResolver` + `WorkflowCatalog` access

This keeps the MCP server decoupled from the resolution mechanism.

## CLI

New and modified commands:

```bash
# Create with MCP enabled
flux service create billing --namespace billing --mcp

# New: update service-level settings
flux service update billing --mcp
flux service update billing --no-mcp

# Start with stored MCP setting (or override)
flux service start billing --port 9000
flux service start billing --port 9000 --mcp
flux service start billing --port 9000 --no-mcp

# Show includes MCP status
flux service show billing
#   MCP: enabled
#   MCP URL: /services/billing/mcp/tools
```

### Modified Endpoints

`POST /services` and `PUT /services/{name}` accept `mcp_enabled` (boolean, validated).

`GET /services/{name}` response now includes:
- `mcp_enabled` field
- Per-endpoint `input_schema` (Pydantic JSON Schema or null) and `description` (docstring)

`GET /services` includes `mcp_enabled` per service.

### New Endpoint

```
GET /services/{name}/mcp/tools    → MCP tool info (requires mcp_enabled=true, else 404)
```

Returns:
```json
{
  "service": "billing",
  "mcp_enabled": true,
  "mcp_url": "/services/billing/mcp/tools",
  "tools": [
    {"name": "invoice", "description": "Run invoice synchronously. Process an invoice."},
    {"name": "invoice_async", "description": "Run invoice asynchronously."},
    ...
  ],
  "tool_count": 10
}
```

## New Files

| File | Purpose |
|---|---|
| `flux/service_mcp.py` | `ServiceMCPServer`, `ProxyBackedMCPServer`, `EndpointProvider` protocol, `EndpointInfo`, tool generation with typed/generic parameters |
| `tests/flux/test_service_mcp.py` | 5 unit tests — tool generation, typed params, descriptions |
| `tests/flux/test_schema_extraction.py` | 7 unit tests — Pydantic schema extraction, docstring extraction, edge cases |
| `tests/e2e/fixtures/service_mcp_workflow.py` | Fixture workflows: `typed_greet` (Pydantic input) + `untyped_add` (dict input) |
| `tests/e2e/test_service_mcp.py` | 5 E2E tests — MCP flag lifecycle, schema in endpoints, tool info endpoint |

## Modified Files

| File | Change |
|---|---|
| `flux/models.py` | `mcp_enabled` Boolean column on `ServiceModel` |
| `flux/service_store.py` | `mcp_enabled` on `ServiceInfo`, create/update support |
| `flux/catalogs.py` | `extract_workflow_input_schema()` + `extract_workflow_description()` at registration, temp file cleanup |
| `flux/cli_service.py` | `--mcp` on create, `flux service update` command, `--mcp/--no-mcp` on start, MCP status in show |
| `flux/server.py` | `mcp_enabled`/`input_schema`/`description` in CRUD responses, MCP tools info endpoint, boolean validation |
| `flux/service_proxy.py` | MCP app mount on standalone, startup error handling |
| `docs/advanced-features/services.md` | MCP Integration section + Upgrade Notes |

## Validation

- **50 unit tests pass** (28 store + 10 resolver + 5 MCP tool gen + 7 schema extraction)
- **5 E2E tests pass** (MCP flag, toggle, schema, disabled → 404, tool info endpoint)
- **All existing tests unaffected** (CI green: lint, unit, E2E)

## Upgrade Notes

This release adds the `mcp_enabled` column to the `services` table. Existing SQLite deployments must recreate their database. There is no automatic migration.

### CODE REVIEW

File: flux/server.py [hunk 1]
```diff
@@ -3010,13 +3010,15 @@ async def create_service(
                     namespaces=body.get("namespaces", []),
                     workflows=body.get("workflows", []),
                     exclusions=body.get("exclusions", []),
+                    mcp_enabled=bool(body.get("mcp_enabled", False)),
                 )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3018]: Fixed. Added explicit boolean type check — rejects non-bool values with 400.

File: flux/server.py [hunk 2]
```diff
@@ -3152,13 +3158,15 @@ async def update_service(
                     remove_namespaces=body.get("remove_namespaces"),
                     remove_workflows=body.get("remove_workflows"),
                     remove_exclusions=body.get("remove_exclusions"),
+                    mcp_enabled=body.get("mcp_enabled"),
                 )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3170]: Fixed. Same boolean validation added to update endpoint.

File: flux/service_mcp.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    def _generate_tools(self, endpoints: dict[str, EndpointInfo]) -> None:
+        for key, info in endpoints.items():
+            if key in self._registered_tools:
+                continue
+            self._register_run_tool(info, mode="sync")
+            self._register_run_tool(info, mode="async")
+            self._register_resume_tool(info, mode="sync")
+            self._register_resume_tool(info, mode="async")
+            self._register_status_tool(info)
+            self._registered_tools.add(key)
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_mcp.py:88]: By design — tools are generated per-session. Added comment explaining existing tool names are skipped; schema changes picked up on next client connection.

File: flux/models.py [hunk 4]
```diff
@@ -759,6 +760,7 @@ class ServiceModel(Base):
     namespaces = Column(JSON, nullable=False, default=list)
     workflows = Column(JSON, nullable=False, default=list)
     exclusions = Column(JSON, nullable=False, default=list)
+    mcp_enabled = Column(Boolean, nullable=False, default=False)
     created_at = Column(
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/models.py:764]: Added Upgrade Notes section to docs — existing SQLite deployments must recreate their database.

File: flux/service_proxy.py [hunk 5]
```diff
@@ -107,12 +108,25 @@ def create_standalone_app(
...
+        from flux.service_mcp import create_service_mcp_server
+
+        mcp_server = create_service_mcp_server(service_name, proxy._client)
+
+        @app.on_event("startup")
+        async def _init_mcp_tools():
+            endpoints = await mcp_server.provider.get_endpoints()
+            mcp_server._generate_tools(endpoints)
+
+        app.mount("/mcp", mcp_server.mcp.http_app())
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:142]: Fixed. Startup hook wrapped in try/except with warning log. MCP starts with empty tools and retries on first connection.

File: flux/server.py [hunk 6]
```diff
@@ -3195,6 +3203,79 @@ async def delete_service(
                 logger.error(f"Error deleting service: {str(e)}")
                 raise HTTPException(status_code=500, detail=f"Error deleting service: {str(e)}")
 
+        @api.get("/services/{service_name}/mcp")
+        async def service_mcp_info(
+            service_name: str,
+            identity: FluxIdentity = Depends(get_identity),
+        ):
+            from flux.service_store import ServiceStore
+            from flux.service_resolver import ServiceResolver, CollisionError
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py]: Fixed. Renamed route to /services/{name}/mcp/tools to clarify it's an info endpoint, not MCP transport. Updated docs.

File: flux/catalogs.py [hunk 7]
```diff
@@ -19,6 +19,76 @@
...
+        except ImportError:
+            pass
+
+        return None
+    except Exception:
+        return None
+    finally:
+        sys.modules.pop(mod_name, None)
+
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/catalogs.py:89]: Added trust boundary comment. This runs in the same context as workflow registration — the server already executes workflow source for the catalog.

File: flux/catalogs.py [hunk 8]
```diff
@@ -19,6 +19,76 @@
...
+    import sys
+    import tempfile
+    import typing
+
+    mod_name = f"_schema_extract_{workflow_name}"
+    try:
+        with tempfile.NamedTemporaryFile(suffix=".py", delete=False, mode="wb") as f:
+            f.write(source)
+            f.flush()
+            spec = importlib.util.spec_from_file_location(mod_name, f.name)
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/catalogs.py:46]: Fixed. Temp file is now cleaned up via os.unlink in the finally block.

File: flux/catalogs.py [hunk 9]
```diff
@@ -216,6 +298,12 @@ def parse(self, source: bytes) -> list[WorkflowInfo]:
                             node,
                             tree,
                         )
+                        input_schema = extract_workflow_input_schema(source, node.name)
+                        if input_schema is not None:
+                            wf_metadata["input_schema"] = input_schema
+                        description = extract_workflow_description(source, node.name)
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/catalogs.py]: Fixed. Added `_enrich_workflow_metadata()` that loads the module once for all workflows in the file. Per-workflow extraction calls removed from the loop.

File: flux/server.py [hunk 10]
```diff
@@ -3195,6 +3211,79 @@ async def delete_service(
...
+                                "name": f"status_{name}",
+                                "description": f"Check {name} execution status.",
+                            },
+                        ],
+                    )
+
+                return {
+                    "service": service_name,
+                    "mcp_enabled": True,
+                    "mcp_url": f"/services/{service_name}/mcp/tools",
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py]: Fixed. Renamed `mcp_url` to `tools_url` to clarify it's an info endpoint.

File: flux/server.py [hunk 11]
```diff
@@ -3195,6 +3211,79 @@ async def delete_service(
                 logger.error(f"Error deleting service: {str(e)}")
                 raise HTTPException(status_code=500, detail=f"Error deleting service: {str(e)}")
 
+        @api.get("/services/{service_name}/mcp/tools")
+        async def service_mcp_info(
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/server.py:3215]: PR description already updated to reference `/services/{name}/mcp/tools`.

File: tests/flux/test_service_mcp.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+def _build_server(endpoints: dict[str, EndpointInfo]) -> ServiceMCPServer:
+    provider = FakeProvider(endpoints)
+    server = ServiceMCPServer("test-service", provider)
+    server._generate_tools(endpoints)
+    return server
+
+
+def _tool_names(server: ServiceMCPServer) -> set[str]:
+    return {t.name for t in server.mcp._tool_manager.list_tools()}
+
```
[the remark this answers is not in the document]
Comment (edurdias) [reply tests/flux/test_service_mcp.py:40]: Fixed. Added `tool_names` property and `get_tool_function()` on `ServiceMCPServer`. Tests use these public APIs instead of FastMCP internals.

File: flux/service_proxy.py [hunk 13]
```diff
@@ -107,12 +108,32 @@ def create_standalone_app(
...
+        async def _init_mcp_tools():
+            try:
+                endpoints = await mcp_server.provider.get_endpoints()
+                mcp_server._generate_tools(endpoints)
+            except Exception as e:
+                import logging
+
+                logging.getLogger(__name__).warning(
+                    f"Failed to initialize MCP tools: {e}. MCP will retry on first connection.",
+                )
```
[the remark this answers is not in the document]
Comment (edurdias) [reply flux/service_proxy.py:128]: Fixed. Removed misleading retry message. Now honestly states: MCP endpoint mounted but has no tools, restart once Flux server is available.
